src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  function deepClone(obj) {
+    return JSON.parse(JSON.stringify(obj));
+}
+
+let currentState = deepClone(state);
+const stateHistory = [];
+
+for (const action of actions) {
+    switch (action.type) {
+        case 'clear':
+            currentState = {};
+            break;
+        case 'addProperties':
+            currentState = {
+                ...currentState,
+                ...action.extraData,
+            };
+            break;
+        case 'removeProperties':
+            currentState = { ...currentState };
```

Comment:
Creating a new copy of `currentState` here is redundant since you are already creating a copy in the `addProperties` case. To follow the DRY principle, you can move this operation outside of the `switch` statement.